sdks/java/expansion-service/src/main/java/org/apache/beam/sdk/expansion/service/ExpansionService.java

```diff
@@ -524,4 +554,15 @@ public static void main(String[] args) throws Exception {
     server.start();
     server.awaitTermination();
   }
+
+  private static class NoOpRunner extends PipelineRunner<PipelineResult> {
+    public static NoOpRunner fromOptions(PipelineOptions opts) {
+      return new NoOpRunner();
+    }
+
```

Comment:
Please move NoOpRunner to a separate class and add basic unit tests (pipeline creation with and without options, expansion etc.).

Reply:
The Pipeline in ExpansionService is not supposed to be run. First of all, the Pipeline specification is not finished at that time, so calling run() should result in some sort of error. UnsupportedOperationException seems to be like one of the best options in this sense. That is probably even better than using DirectRunner, which might in that case run somewhat partial Pipeline.
I generally like having tests for everything that makes sense to be tested, but in this case, I'm not sure what to test, because the  `NoOpRunner` (we might come up with a better name, for instance `ExpansionRunner`?) is simply throwing exception and not doing enything else.

Reply:
I'd say, that the expansion process is tested in https://github.com/apache/beam/blob/efc9b7fd8d4f7cedf006911d86b4467f40812046/sdks/java/expansion-service/src/test/java/org/apache/beam/sdk/expansion/service/ExpansionServiceTest.java#L120 
The main point is that the expansion cannot call `run` method, as that really does not make sense in this context.  It would simplify things in the context of ExpansionService, if the runner can be `null`, but that seems to be the case only for ExpansionService, and I'd prefer to keep the validations as they are. Maybe we can rename NoOpRunner to ThrowingRunner? Or NotRunnableRunner?